File: 1052-grumpy-bookstore-owner/1052-grumpy-bookstore-owner.py
class Solution:
    def maxSatisfied(self, customers: List[int], grumpy: List[int], minutes: int) -> int:
        n = len(customers)
        total = 0
        dp = [0]
        
        for c, g in zip(customers, grumpy):
            if g == 0:
                total += c
                dp.append(dp[-1])
                
            if g == 1:
                dp.append(dp[-1] + c)
        _max = 0    
        for i in range(minutes, n + 1):
            _max = max(_max, dp[i] - dp[i - minutes])
            
        return total + _max
        

File: 1052-grumpy-bookstore-owner/1052-grumpy-bookstore-owner.rs
impl Solution {
    pub fn max_satisfied(customers: Vec<i32>, grumpy: Vec<i32>, minutes: i32) -> i32 {
        let n = customers.len();
        let mut total = 0;
        let mut dp: Vec<i32> = vec![0];
        let mut max = 0;
        let minutes = minutes as usize;
        
        for (c, g) in customers.iter().zip(&grumpy) {
            let m = dp.len() - 1;
            if *g == 0 {
                total += *c;
                dp.push(dp[m]);
            }
            
            if *g == 1 {
                dp.push(dp[m] + *c);
            }
            
            if m + 1 >= minutes {
                max = std::cmp::max(max, dp[m + 1] - dp[m + 1 - minutes]);
            }
        }
            
        total + max
    }
}